### PR TITLE
Pawn promotion

### PR DESCRIPTION
**Pull request by Andrew-Lipson**

### CODE REVIEW

File: src/main/java/Controller/Controller.java [hunk 1]
```diff
@@ -48,6 +49,12 @@ public void update() {
         mainview.updateView(game.getCompleteFEN());
     }
 
+    @Override
+    public void pawnPromotion() {
+        update();
+        mainview.promotionPopup(game.getWhitesTurn());
```
Comment (ryan-saffer) [inline src/main/java/Controller/Controller.java:55]: how about `displayPromotionPopup()`

File: src/main/java/Controller/Controller.java [hunk 2]
```diff
@@ -77,4 +84,10 @@ public void handleCircleClicked(PositionView positionView) {
         game.makeAMove(previousPosition,newPosition);
         this.clickedPiece = null;
     }
+
+    @Override
+    public void handlePromotionClicked(String string) {
+        game.promotionPieceDecision(PieceType.getPieceType(string));
```
Comment (ryan-saffer) [inline src/main/java/Controller/Controller.java:90]: `game.onPromotionPieceDecided()`

File: src/main/java/Model/Game.java [hunk 3]
```diff
@@ -114,11 +115,14 @@ public void makeAMove(Position previousPosition, Position newPosition) {
...
-        endTurn();
-        updateView();
-
+        disableCastlingIfRequired(piece, previousPosition);
+        if(canPromote(newPosition, piece)) {
+            promotionPiece = piece;
+            _observer.pawnPromotion();
+        } else {
+            endTurn();
+            updateView();
```
Comment (ryan-saffer) [inline src/main/java/Model/Game.java:124]: could place `updateView()` inside `endTurn()`

File: src/main/java/View/MainView.java [hunk 4]
```diff
@@ -145,6 +147,10 @@ public void removeMoveOptionsCircles() {
         }
     }
 
+    public void promotionPopup(boolean isWhite){
+        new PromotionView(listener, isWhite);
```
Comment (ryan-saffer) [inline src/main/java/View/MainView.java]: same feedback as Controller class:
```java
PromotionView promotionView = new PromotionView(listener, isWhite);
promotionView.show()
```

File: src/main/java/View/PromotionView.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                yOffset = event.getSceneY();
+            }
+        });
+        root.setOnMouseDragged(new EventHandler<MouseEvent>() {
+            @Override
+            public void handle(MouseEvent event) {
+                promotionStage.setX(event.getScreenX() - xOffset);
+                promotionStage.setY(event.getScreenY() - yOffset);
+            }
+        });
```
Comment (ryan-saffer) [inline src/main/java/View/PromotionView.java]: can be simplified with lambda:
```java
root.setOnMousePressed(event -> {
    xOffset = event.getSceneX();
    yOffset = event.getSceneY();
});
root.setOnMouseDragged(event -> {
    promotionStage.setX(event.getScreenX() - xOffset);
    promotionStage.setY(event.getScreenY() - yOffset);
});
```

File: src/main/java/View/PromotionView.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            case "b":
+                output += "Bishop";
+                break;
+        }
+
+        output += "-";
+        output += isWhite ? "White" : "Black";
+        output+=".png";
+        return output;
+    }
```
Comment (ryan-saffer) [inline src/main/java/View/PromotionView.java]: i think this could be shared between the views - dont you have similar code somewhere else?

**Review by ryan-saffer — APPROVED** [review]

File: src/main/java/Model/Game.java (hunk 3 above)
Comment (Andrew-Lipson) [reply src/main/java/Model/Game.java:124]: I like having them separate. `endTurn()` is for game logic where as `updateView()` calls the observer

File: src/main/java/View/PromotionView.java (hunk 6 above)
Comment (Andrew-Lipson) [reply src/main/java/View/PromotionView.java]: Would this be a good time to use a static method? It's in SquareView but PromotionView has no reason to have a SquareView instance

**Pull request merged**